polkadot/collator/src/lib.rs

```diff
@@ -117,30 +133,148 @@ pub fn collate_ingress<'a, R>(relay_context: R)
 		})
 		.map(|ordered| ordered.into_iter().map(|((_, id), egress)| (id, egress)))
 		.map(|i| i.collect::<Vec<_>>())
-		.map(ConsolidatedIngress))
+		.map(ConsolidatedIngress)
 }
 
-/// Produce a candidate for the parachain.
-pub fn collate<'a, R, P>(local_id: ParaId, relay_context: R, para_context: P)
-	-> Box<Future<Item=parachain::Candidate, Error=R::Error> + 'a>
+/// Produce a candidate for the parachain, with given contexts, parent head, and signing key.
+pub fn collate<'a, R, P>(
+	local_id: ParaId,
+	last_head: HeadData,
+	relay_context: R,
+	para_context: P,
+	key: Arc<ed25519::Pair>,
+)
+	-> impl Future<Item=parachain::Collation, Error=R::Error> + 'a
 	where
-		R: RelayChainContext,
-	    R::Error: 'a,
+		R: RelayChainContext + 'a,
+		R::Error: 'a,
 		R::FutureEgress: 'a,
 		P: ParachainContext + 'a,
 {
-	Box::new(collate_ingress(relay_context).map(move |ingress| {
-		let (block_data, _, signature) = para_context.produce_candidate(
+	collate_ingress(relay_context).map(move |ingress| {
+		let (block_data, head_data) = para_context.produce_candidate(
+			last_head,
 			ingress.0.iter().flat_map(|&(id, ref msgs)| msgs.iter().cloned().map(move |msg| (id, msg)))
 		);
 
-		parachain::Candidate {
+		let signature = key.sign(&block_data.0[..]).into();
+		let pubkey_bytes: [u8; 32] = key.public().into();
+
+		let receipt = parachain::CandidateReceipt {
 			parachain_index: local_id,
-			collator_signature: signature,
-			block: block_data,
-			unprocessed_ingress: ingress,
+			collator: pubkey_bytes.into(),
+			signature,
+			head_data,
+			balance_uploads: Vec::new(),
+			egress_queue_roots: Vec::new(),
+			fees: 0,
+		};
+
+		parachain::Collation {
+			receipt,
+			block_data,
+		}
+	})
+}
+
+/// Polkadot-api context.
+struct ApiContext;
+
+impl RelayChainContext for ApiContext {
+	type Error = ();
+	type FutureEgress = Result<Vec<Vec<Message>>, Self::Error>;
+
+	fn routing_parachains(&self) -> BTreeSet<ParaId> {
+		BTreeSet::new()
+	}
+
+	fn unrouted_egress(&self, _id: ParaId) -> Self::FutureEgress {
+		Ok(Vec::new())
+	}
+}
+
+struct CollationNode<P, E> {
+	parachain_context: P,
+	exit: Option<E>,
+	para_id: ParaId,
+	key: Arc<ed25519::Pair>,
+}
+
+impl<P, E> Worker for CollationNode<P, E> where
+	P: ParachainContext + 'static,
+	E: Future<Item=(),Error=()> + Send + 'static
+{
+	type Work = Box<Future<Item=(),Error=()>>;
+	type Exit = future::Either<E, future::Empty<(),()>>;
+
+	fn exit(&mut self) -> Self::Exit {
+		match self.exit.take() {
+			Some(exit) => future::Either::A(exit),
+			None => future::Either::B(future::empty()),
```

Comment:
isn't this the return value that is then passed into the separate thread, which then performs an `exit.wait()` before exiting? With using `future::empty()`  this would prevent the process from exiting ever, right? Is that intentional? If so, why do we want to prevent an exit? Or did you mean to use `future::ok()` to have it resolve immediately, which would make more sense to me...